File: testdata_loader.go
package mowgli

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
)

// TestCase represents a single test case
type TestCase struct {
	Name          string         `json:"name"`
	Data          map[string]any `json:"data"`
	ExpectedValid bool           `json:"expectedValid"`
}

// TestCaseFile represents a file containing multiple test cases
type TestCaseFile struct {
	TestCases []TestCase `json:"testCases"`
}

// LoadSpec loads a spec JSON file from the testdata directory
func LoadSpec(filename string) (*Spec, error) {
	path := filepath.Join("testdata", "specs", filename)
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("failed to read spec file %s: %w", path, err)
	}

	spec, err := ParseSpecString(string(data))
	if err != nil {
		return nil, fmt.Errorf("failed to parse spec from %s: %w", path, err)
	}

	return spec, nil
}

// LoadTestCases loads test cases from a JSON file in the testdata directory
func LoadTestCases(filename string) ([]TestCase, error) {
	path := filepath.Join("testdata", "cases", filename)
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("failed to read test case file %s: %w", path, err)
	}

	var testFile TestCaseFile
	if err := json.Unmarshal(data, &testFile); err != nil {
		return nil, fmt.Errorf("failed to parse test case file %s: %w", path, err)
	}

	return testFile.TestCases, nil
}

File: js/testdata-loader.ts
import { readFileSync } from 'fs';
import { join } from 'path';
import { parseSpec, validate, Spec } from './src/index';

export interface TestCase {
  name: string;
  data: Record<string, any>;
  expectedValid: boolean;
}

export interface TestCaseFile {
  testCases: TestCase[];
}

/**
 * Loads a spec JSON file from the testdata directory
 */
export function loadSpec(filename: string): Spec {
  const path = join(__dirname, '..', 'testdata', 'specs', filename);
  const data = readFileSync(path, 'utf-8');
  return parseSpec(data);
}

/**
 * Loads test cases from a JSON file in the testdata directory
 */
export function loadTestCases(filename: string): TestCase[] {
  const path = join(__dirname, '..', 'testdata', 'cases', filename);
  const data = readFileSync(path, 'utf-8');
  const testFile: TestCaseFile = JSON.parse(data);
  return testFile.testCases;
}

/**
 * Runs all test cases against a spec and returns results
 */
export function runTestCases(spec: Spec, testCases: TestCase[]): {
  passed: number;
  failed: number;
  failures: Array<{ name: string; expected: boolean; actual: boolean; errors?: any[] }>;
} {
  let passed = 0;
  let failed = 0;
  const failures: Array<{ name: string; expected: boolean; actual: boolean; errors?: any[] }> = [];

  for (const testCase of testCases) {
    const result = validate(testCase.data, spec);
    if (result.valid === testCase.expectedValid) {
      passed++;
    } else {
      failed++;
      failures.push({
        name: testCase.name,
        expected: testCase.expectedValid,
        actual: result.valid,
        errors: result.errors,
      });
    }
  }

  return { passed, failed, failures };
}

